torch/linalg/__init__.py

```diff
@@ -13,3 +15,248 @@
 
 Alias of :func:`torch.ger`.
 """)
+
+@overload  # noqa: 749
+def norm(input, ord=None, dim=None, keepdim=False, out=None, dtype=None):  # noqa: 749
+    # type: (Tensor, str, Optional[List[int]], bool, Optional[Tensor], Optional[int]) -> Tensor
+    pass
+
+@overload  # noqa: 749
+def norm(input, ord=None, dim=None, keepdim=False, out=None, dtype=None):  # noqa: 749
+    # type: (Tensor, Optional[number], Optional[List[int]], bool, Optional[Tensor], Optional[int]) -> Tensor
+    pass
+
+@overload  # noqa: 749
+def norm(input, ord=None, dim=None, keepdim=False, out=None, dtype=None):  # noqa: 749
+    # type: (Tensor, Optional[number], Optional[int], bool, Optional[Tensor], Optional[int]) -> Tensor
+    pass
+
+@overload  # noqa: 749
+def norm(input, ord=None, dim=None, keepdim=False, out=None, dtype=None):  # noqa: 749
+    # type: (Tensor, str, Optional[int], bool, Optional[Tensor], Optional[int]) -> Tensor
+    pass
+
+def norm(input, ord=None, dim=None, keepdim=False, out=None, dtype=None) -> Tensor:  # noqa: 749
```

Comment:
Would it be better to make this a C++ function? I'm not 100% sure if string arguments in `native_functions.yaml` schemas are supported. I don't see any examples of that.

Reply:
They are:

https://github.com/pytorch/pytorch/blob/896319a17fda0642a91784938dd63fb25d1345e5/aten/src/ATen/native/native_functions.yaml#L1127

Reply:
C++ would make this a lot simpler.

Reply:
Ah, thanks! I'll go ahead and do that then.

Reply:
While moving this function over to C++, I ran into an issue with trying to make the `dim` argument support `None`, int, and tuple values properly. I ran into this issue before (it blocked https://github.com/pytorch/pytorch/issues/29137#issuecomment-578538543), and as far as I know, the issue hasn't been fixed yet.

Basically, the way codegen currently works, `dim=None` gets turned into `dim=()` if the schema has `dim` as a kwarg with type `int[1]`. But for numpy compatibility, `dim=None` and `dim=()` have to be treated as separate cases.

The last time I ran into this issue, I was very new to PyTorch and didn't understand the situation as much as I do now. So I tried to work around it this time, in a few different ways. But these did not work for various reasons:
* Overload schemas with `int[]? dim` and `int dim`. Results in ambiguous C++ function calls
* Overload schemas with `int[1] dim` and `int? dim`. Results in ambiguous C++ function calls
* Overload schemas with `int[1] dim` in one set, remove the `dim` argument from the other set and dispatch it to a function with a different name (`linalg_norm_no_dim`, rather than `linalg_norm`). Results in ambiguous C++ function calls because evidently the dispatch function name only relates to the `at::native::` namespace--an `at::linalg_norm` function (not `at::linalg_norm_no_dim`) is apparently automatically generated for each schema.

I also tried changing `dim` to be a positional `int[1]` arg, rather than a kwarg, with a set of overloads that don't have the `dim` argument. This is what `torch.sum` does, and it at least builds and avoids the issue of `dim=None` being silently converted to `dim=()`. However, this makes it impossible to call `torch.linalg.norm(..., dim=None, ...)`. In order to get `dim` to be `None`, it has to be excluded from the call arguments. While not ideal, this seems like the best supported way to set up the schemas for this function, as far as I can tell.

If we do go this route, we'll have to have 8 schemas for `linalg_norm`. Starting with the 2 schemas required to work around this issue, that gets multiplied by 2 to support `out`, and multiplied by 2 again to support `p=None`, `p=int`, and `p=str`. However, I haven't fully tested these schemas yet, so hopefully I don't find another issue that splits it further into 16.

Here are the schemas I landed on:
```
- func: linalg_norm(Tensor self, Scalar? ord=None, *, bool keepdim=False, ScalarType? dtype=None) -> Tensor
  variants: function, method

- func: linalg_norm.ord_str(Tensor self, str ord, *, bool keepdim=False, ScalarType? dtype=None) -> Tensor
  variants: function, method

- func: linalg_norm.out(Tensor self, Scalar? ord=None, *, bool keepdim=False, ScalarType? dtype=None, Tensor(a!) out) -> Tensor(a!)
  variants: function

- func: linalg_norm.ord_str_out(Tensor self, str ord, *, bool keepdim=False, ScalarType? dtype=None, Tensor(a!) out) -> Tensor(a!)
  variants: function

- func: linalg_norm.dim(Tensor self, Scalar? ord, int[1] dim, *, bool keepdim=False, ScalarType? dtype=None) -> Tensor
  variants: function, method

- func: linalg_norm.ord_str_dim(Tensor self, str ord, int[1] dim, *, bool keepdim=False, ScalarType? dtype=None) -> Tensor
  variants: function, method

- func: linalg_norm.dim_out(Tensor self, Scalar? ord, int[1] dim, *, bool keepdim=False, ScalarType? dtype=None, Tensor(a!) out) -> Tensor(a!)
  variants: function

- func: linalg_norm.ord_str_dim_out(Tensor self, str ord, int[1] dim, *, bool keepdim=False, ScalarType? dtype=None, Tensor(a!) out) -> Tensor(a!)
  variants: function
```

@mruberry , what's your take on this? Do you think this is a sufficient solution for now?

Reply:
I would really like to avoid your having to do that and bloating the code with it. @bhosmer, what are your thoughts? Would a custom binding, like the one we collaborated on for torch.full, work in this case?

Reply:
I just realized that since the above schemas have `dim` as a positional arg, it can't be called like `torch.linalg_norm(tensor, dim=dim)`--the `ord` argument has to be included. So even despite code bloat, I don't think this is a viable solution.

Reply:
@mruberry , are you talking about the dispatch code related to `full` in [tools/autograd/templates/python_torch_functions.cpp](https://github.com/pytorch/pytorch/blob/master/tools/autograd/templates/python_torch_functions.cpp#L236). I could do something similar to that for `norm`. Should I go ahead and try it? Because as far as I can tell, there's currently no other way to get the behavior I want.

Reply:
That's what I was thinking. Yes. Let me ping @bhosmer again for his thinking.

Reply:
I had a chance to speak with @bhosmer, and I wanted to recap some of our thinking and be sure we're on the same page, @kurtamohler. Let's consider signatures like this:

```
- func: linalg_norm.ord_scalar(Tensor self, Scalar? ord=None, *, ScalarType? dtype=None) -> Tensor
- func: linalg_norm.ord_str(Tensor self, str ord=None, *, ScalarType? dtype=None) -> Tensor
- func: linalg_norm.ord_out(Tensor self, Scalar? ord=None, *, ScalarType? dtype=None, Tensor(a!) out) -> Tensor(a!)
- func: linalg_norm.ord_str_out(Tensor self, str ord=None, *, ScalarType? dtype=None, Tensor(a!) out) -> Tensor(a!)
- func: linalg_norm.ord_dim(Tensor self, Scalar ord=None, int[1] dim, bool keepdim=False *, ScalarType? dtype=None) -> Tensor
- func: linalg_norm.ord_str_dim(Tensor self, str ord=None, int[1] dim, book keepdim=False *, ScalarType? dtype=None) -> Tensor
- func: linalg_norm.ord_dim_out(Tensor self, Scalar ord=None, int[1] dim, bool keepdim=False *, ScalarType? dtype=None, Tensor(a!) out) -> Tensor(a!)
- func: linalg_norm.ord_str_dim_out(Tensor self, str ord=None, int[1] dim, book keepdim=False *, ScalarType? dtype=None, Tensor(a!) out) -> Tensor(a!)
```

It seems like these signatures have two issues. First, if there weren't the dim ambiguity issue there would be only 4 signatures:

```
- func: linalg_norm.ord_scalar(Tensor self, Scalar? ord=None, int[1]? dim, bool keepdim=False, *, ScalarType? dtype=None) -> Tensor
- func: linalg_norm.ord_str(Tensor self, str ord=None, int[1]? dim, bool keepdim=False, *, ScalarType? dtype=None) -> Tensor
- func: linalg_norm.ord_out(Tensor self, Scalar? ord=None, int[1]? dim, bool keepdim=False, *, ScalarType? dtype=None, Tensor(a!) out) -> Tensor(a!)
- func: linalg_norm.ord_str_out(Tensor self, str ord=None, int[1]? dim, bool keepdim=False, *, ScalarType? dtype=None, Tensor(a!) out) -> Tensor(a!)
```

which is more compact. Second, in Python you can specify the dim argument without specifying the ord argument. That is, in Python you can write:

```
def fn(x, one=1, two=2):
  return x + one + two
fn(0, two=4)
```

Skipping the specification of one. NumPy, since its functions are defined in Python, permits the same:

```
def norm(x, ord=None, axis=None, keepdims=False):
  ...

np.linalg.norm(np.array((1, 2, 3)), axis=0)
```

But `torch.norm`, as written, would not allow this.

Is this an accurate summary of the two issues? If it is, maybe both of these problems are OK for now? The first isn't user-facing, and for the second we could write a Python binding (in Python) if we really wanted to. That seems separable, however.

Reply:
Hi @kurtamohler @mruberry - did some digging and without writing out the full justification (just lateness), I think we have two options:

* immediate: the way to stay sane given current schema support is just to use `int[]? dim`, and forego the bare-int form (`dim=5`). I believe this type works as expected (i.e., passes `None` when the arg is missing, etc.), and as Kurt experienced, trying to finesse the situation by splitting responsibility between two signatures (or rather, pairs of signatures) runs into various other constraints.

* alternatively, I've got a PR (#43262) that adds support for filled optional int lists (e.g. `int[2]?`). Assuming I haven't forgotten about any dealbreakers around doing this (we'd discussed it previously, but IIRC it wasn't done for reasons of priority), that should give you what you need.

Reply:
I think the idea of just using `int[]? dim` with only 4 schemas makes sense for right now, so I'll go ahead and do that. I'm happy to change it to `int[n]?` once support for that is merged, thanks for doing that @bhosmer!

Reply:
Alright, I implemented `at::linalg_norm` with the schemas we talked about and removed the Python implementation. I also realized that the code I copied and modified from `torch.functional.norm` was overly complicated, so I was able to simplify it quite a bit.

I haven't enabled tests in `common_methods_invocations.py` yet, so I'll do that next.